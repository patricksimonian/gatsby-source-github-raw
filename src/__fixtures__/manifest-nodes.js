export const MANIFEST_SOURCE_TYPE = 'foo';
const NON_MANIFEST_SOURCE_TYPE = 'baz';

export const GRAPHQL_NODES_WITHOUT_MANIFEST_JSON = [
  {
    id: '56bcf513-818b-5188-99e9-e194b32eaf1a',
    internal: {
      type: NON_MANIFEST_SOURCE_TYPE,
    },
  },
  {
    id: '56bcf513-818b-5188-99e9-e194b32eaf1a',
    internal: {
      type: NON_MANIFEST_SOURCE_TYPE,
    },
  },
  {
    id: '56bcf513-818b-5188-99e9-e194b32eaf1a',
    internal: {
      type: NON_MANIFEST_SOURCE_TYPE,
    },
  },
];

export const GRAPHQL_NODES_WITH_MANIFEST_JSON = [
  {
    id: '56bcf513-818b-5188-99e9-e194b32eaf1a',
    internal: {
      type: MANIFEST_SOURCE_TYPE,
    },
  },

  {
    id: '62c295c4-7df2-556c-a8d0-0a49a3761ca1',
    internal: {
      type: MANIFEST_SOURCE_TYPE,
    },
  },
  {
    id: '0597ed6f-b09f-5d0a-b9b1-16ac7d3a3ca4',
    internal: {
      type: MANIFEST_SOURCE_TYPE,
    },
  },
].concat(GRAPHQL_NODES_WITHOUT_MANIFEST_JSON);
