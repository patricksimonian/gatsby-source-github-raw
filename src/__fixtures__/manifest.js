export const MANIFEST = [
  {
    url: 'https://github.com/foo/bar/blob/master/foo.md',
  },
  {
    url: 'https://github.com/foo/bar/blob/master/bar.md',
    personas: ['Developer', 'Designer'],
    labels: ['foo', 'bar'],
  },
  {
    url: 'https://github.com/foo/bar/blob/master/baz.md',
  },
];
