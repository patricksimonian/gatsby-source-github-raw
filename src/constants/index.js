export { NOMENCLATURE, ERRORS } from './strings';
export { JSON_SCHEMA } from './schema';
