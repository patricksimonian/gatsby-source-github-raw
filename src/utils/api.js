/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
import fetch from 'node-fetch';
/**
 * Fetches contents from file
 * note the media type header, it converts what would have been a
 * b64 encoded string of the file contents into either raw string data or json
 * @param {String} path
 * @param {String} token
 * @param {Object} boundProperties and object properties you want to bind to returned json
 */
export const fetchFile = async (path, token) => {
  try {
    const result = await fetch(path, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`,
        'X-GitHub-Media-Type': 'Accept: application/vnd.github.v3.raw+json',
      },
    });
    const data = await result.json();
    if (result.ok) return data;
  } catch (e) {
    return undefined;
  }
  return undefined;
};

/**
 * returns array of fetch file promises
 * @param {Array} files array of github api contents api uri strings
 * @param {*} token github token
 */
export const fetchFiles = (files, token) => {
  return files.map(f => fetchFile(f.path, token));
};
