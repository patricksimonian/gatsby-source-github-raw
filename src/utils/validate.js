/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
import validUrl from 'valid-url';
import Url from 'url';
import { isFunction, isString, isArray, isPlainObject } from 'lodash';
import { NOMENCLATURE } from '../constants';
export const getLodashValidateFnByMapping = mapping => {
  switch (mapping) {
    case 'String':
      return isString;
    case 'Function':
      return isFunction;
    default:
      return () => false;
  }
};
/**
 * validates an object against a schema
 * schema is in format
 * {
 *  [object key name] : {
 *    type: [object constructor String | Array | Object etc],
 *    required: [Boolean]
 *  }
 * }
 * @param {Object} obj the object that is being tested
 * @param {Object} schema the schema object that is being tested against
 * @returns {Object} an object containing error messages and isValid property
 * {
 *   messages: {Array},
 *   isValid: {Boolean}
 * }
 */
export const validateAgainstSchema = (obj, schema) => {
  const error = {
    messages: [],
  };

  Object.keys(schema).every(key => {
    const schemaItem = schema[key];
    let isValid = true;
    // is there a validator function
    if (isFunction(schemaItem.validate)) {
      isValid = schemaItem.validate(obj);
    } else if (schemaItem.required) {
      isValid =
        Object.prototype.hasOwnProperty.call(obj, key) &&
        getLodashValidateFnByMapping(schemaItem.type)(obj[key]);
      // does this source property have it anyways?
    } else if (Object.prototype.hasOwnProperty.call(obj, key)) {
      isValid = getLodashValidateFnByMapping(schemaItem.type)(obj[key]);
    }

    if (!isValid) {
      error.messages.push(
        `Error Validating ${NOMENCLATURE.manifest}: failed on property ${key}, received value ${
          obj[key]
        }`,
      );
    }
    return isValid;
  });

  error.isValid = error.messages.length === 0;

  return error;
};

export const isAGithubUrl = url => {
  if (validUrl.isWebUri(url)) {
    const parsed = new Url.URL(url);
    return parsed.host === 'github.com';
  }
  return false;
};

/**
 * validates source plugin options
 * @param {String} token the github access token
 * @param {String | Array} manifest the manifest config or the node internal.type that points to all the manifest files
 */
export const areOptionsOkay = (token, manifest) =>
  isString(token) && (isString(manifest) || (isArray(manifest) && manifest.every(isPlainObject)));
