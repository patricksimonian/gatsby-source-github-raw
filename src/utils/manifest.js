/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
import { NOMENCLATURE, JSON_SCHEMA } from '../constants';
import { validateAgainstSchema } from './validate';
/**
 * Filter out all nodes to get the ones specify for registry yaml file
 * @param {Function} getNodes gatsby builtin function to return all nodes
 * @param {String} manifestSourceType the internal type refering to registry yaml source
 * @returns {Array} the list of manifest items
 */
export const getManifestInFileSystem = (getNodes, manifestSourceType) => {
  const manifest = getNodes().filter(node => node.internal.type === manifestSourceType);
  if (manifest.length > 0) {
    // filter out internal so that this manifest is exactly the same in structure as a json object
    // being passed in as config
    // eslint-disable-next-line no-unused-vars
    return manifest.map(({ internal, id, parent, children, ...rest }) => rest);
  }

  throw new Error(`${NOMENCLATURE.manifest} not found`);
};

/**
 * @param {Object} manifestItem the registry item found within the registry file sources[index]
 * @returns {Object} {isValid: {Boolean}, messages: []{String} }
 */
export const validateManifestItem = manifestItem =>
  validateAgainstSchema(manifestItem, JSON_SCHEMA);

/**
 * filters out unvalidated manifest items
 * @param {Array} manifest
 * @returns {Array} a new array of filtered manifests
 */
export const validateAndFilterManifest = manifest =>
  manifest.filter(item => {
    const validated = validateManifestItem(item);
    return validated.isValid;
  });
