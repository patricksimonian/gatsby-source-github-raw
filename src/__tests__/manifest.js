import { getManifestInFileSystem } from '../utils';
import { validateAgainstSchema } from '../utils/validate';
import { NOMENCLATURE } from '../constants';
import {
  GRAPHQL_NODES_WITH_MANIFEST_JSON,
  GRAPHQL_NODES_WITHOUT_MANIFEST_JSON,
  MANIFEST_SOURCE_TYPE,
} from '../__fixtures__/manifest-nodes';
import { validateAndFilterManifest, seperateUrlFromMetadata } from '../utils/manifest';
import { MANIFEST } from '../__fixtures__/manifest';

jest.mock('../utils/validate.js');

describe('Manifest Utilities', () => {
  describe('getManifestInFileSystem', () => {
    it('returns the manifest', () => {
      const getNodes = jest.fn(() => GRAPHQL_NODES_WITH_MANIFEST_JSON);
      const manifest = getManifestInFileSystem(getNodes, MANIFEST_SOURCE_TYPE);
      expect(manifest).toMatchSnapshot();
    });

    it('throws if no manifest exists', () => {
      const getNodes = jest.fn(() => GRAPHQL_NODES_WITHOUT_MANIFEST_JSON);
      expect(() => getManifestInFileSystem(getNodes, MANIFEST_SOURCE_TYPE)).toThrow(
        `${NOMENCLATURE.manifest} not found`,
      );
    });
  });

  describe('validateAndFilterManifest', () => {
    it('filters out manifest items that are invalid', () => {
      validateAgainstSchema
        .mockReturnValueOnce({ isValid: false })
        .mockReturnValue({ isValid: true });

      const newManifest = validateAndFilterManifest(MANIFEST);
      expect(newManifest.length).toBeLessThan(MANIFEST.length);

      expect(newManifest).toMatchSnapshot();
    });
  });

  describe('seperateUrlFromMetadata', () => {
    it('seperates metadata', () => {
      const manifestItem = {
        url: 'foo',
        internal: 'bar',
        data1: true,
      };

      const expected = ['foo', { data1: true }];

      expect(seperateUrlFromMetadata(manifestItem)).toEqual(expected);
    });
  });
});
