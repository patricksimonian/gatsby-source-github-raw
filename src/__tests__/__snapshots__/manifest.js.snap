// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manifest Utilities getManifestInFileSystem returns the manifest 1`] = `
Array [
  Object {
    "id": "56bcf513-818b-5188-99e9-e194b32eaf1a",
    "internal": Object {
      "type": "foo",
    },
    "url": "https://github.com/foo/bar/blob/master/foo.md",
  },
  Object {
    "id": "62c295c4-7df2-556c-a8d0-0a49a3761ca1",
    "internal": Object {
      "type": "foo",
    },
    "url": "https://github.com/foo/bar/blob/master/bar.md",
  },
  Object {
    "id": "0597ed6f-b09f-5d0a-b9b1-16ac7d3a3ca4",
    "internal": Object {
      "type": "foo",
    },
    "url": "https://github.com/foo/bar/blob/master/baz.md",
  },
]
`;

exports[`Manifest Utilities validateAndFilterManifest filters out manifest items that are invalid 1`] = `
Array [
  Object {
    "labels": Array [
      "foo",
      "bar",
    ],
    "personas": Array [
      "Developer",
      "Designer",
    ],
    "url": "https://github.com/foo/bar/blob/master/bar.md",
  },
  Object {
    "url": "https://github.com/foo/bar/blob/master/baz.md",
  },
]
`;
